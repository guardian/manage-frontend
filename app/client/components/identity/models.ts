export enum Theme {
	news = 'news',
	opinion = 'opinion',
	features = 'features',
	sport = 'sport',
	culture = 'culture',
	lifestyle = 'lifestyle',
	comment = 'comment',
	work = 'work',
	fromThePapers = 'From the papers',
}

export enum ErrorTypes {
	GENERAL = 'GENERAL',
	NOT_FOUND = 'NOT_FOUND',
	VALIDATION = 'VALIDATION',
}

export enum ConsentOptionType {
	EMAIL = 'EMAIL',
	NEWSLETTER = 'NEWSLETTER',
	OPT_OUT = 'OPT_OUT',
	SUPPORT_REMINDER = 'SUPPORT_REMINDER',
}

export interface User {
	id: string;
	primaryEmailAddress: string;
	consents: string[];
	username: string;
	validated: boolean;
	title: string;
	firstName: string;
	secondName: string;
	address1: string;
	address2: string;
	address3: string;
	address4: string;
	postcode: string;
	country: string;
	countryCode: string;
	localNumber: number;
}

export interface UserError {
	type: ErrorTypes.VALIDATION;
	error: {
		username: string;
	};
}

export interface UserCollection {
	getCurrentUser: () => Promise<User>;
	save: (user: User) => Promise<User>;
	saveChanges: (original: User, changed: User) => Promise<User>;
	getChangedFields: (original: User, changed: User) => Partial<User>;
}

export interface ConsentOption {
	id: string;
	description?: string;
	frequency?: string;
	name: string;
	theme?: string;
	type: ConsentOptionType;
	subscribed: boolean;
	identityName?: string;
	isProduct?: boolean;
	isChannel?: boolean;
	group?: string;
}

export interface ConsentOptionCollection {
	getAll: () => Promise<ConsentOption[]>;
	subscribe: (option: ConsentOption) => Promise<void>;
	unsubscribe: (option: ConsentOption) => Promise<void>;
	newsletters: (options: ConsentOption[]) => ConsentOption[];
	consents: (options: ConsentOption[]) => ConsentOption[];
	unsubscribeAll: () => Promise<void>;
	findById: (
		options: ConsentOption[],
		id: ConsentOption['id'],
	) => ConsentOption | undefined;
	findByIds: (options: ConsentOption[], ids: string[]) => ConsentOption[];
}

/*
 * The following models are being used to aid in migration of functionality off of a legacy system. There is
 * currently no standard library or API for these values at the Guardian, when these become available, these hardcoded values should be replaced.
 * Note: COUNTRIES comes from a non-standard list used by the Guardian.
 */

export enum Titles {
	MR = 'Mr',
	MRS = 'Mrs',
	MS = 'Ms',
	MX = 'Mx',
	MISS = 'Miss',
	DR = 'Dr',
	PROF = 'Prof',
	REV = 'Rev',
}

/* Extracted from https://github.com/catamphetamine/libphonenumber-js using:
 * const core = require('libphonenumber-js/core');
 * const meta = require('libphonenumber-js/metadata.min.json');
 * const ccs = (new core.Metadata(meta)).countryCallingCodes();
 * const processed = Object.keys(ccs);
 * console.dir(processed, { 'maxArrayLength': null });
 */

export const PHONE_CALLING_CODES: string[] = [
	'1',
	'7',
	'20',
	'27',
	'30',
	'31',
	'32',
	'33',
	'34',
	'36',
	'39',
	'40',
	'41',
	'43',
	'44',
	'45',
	'46',
	'47',
	'48',
	'49',
	'51',
	'52',
	'53',
	'54',
	'55',
	'56',
	'57',
	'58',
	'60',
	'61',
	'62',
	'63',
	'64',
	'65',
	'66',
	'81',
	'82',
	'84',
	'86',
	'90',
	'91',
	'92',
	'93',
	'94',
	'95',
	'98',
	'211',
	'212',
	'213',
	'216',
	'218',
	'220',
	'221',
	'222',
	'223',
	'224',
	'225',
	'226',
	'227',
	'228',
	'229',
	'230',
	'231',
	'232',
	'233',
	'234',
	'235',
	'236',
	'237',
	'238',
	'239',
	'240',
	'241',
	'242',
	'243',
	'244',
	'245',
	'246',
	'247',
	'248',
	'249',
	'250',
	'251',
	'252',
	'253',
	'254',
	'255',
	'256',
	'257',
	'258',
	'260',
	'261',
	'262',
	'263',
	'264',
	'265',
	'266',
	'267',
	'268',
	'269',
	'290',
	'291',
	'297',
	'298',
	'299',
	'350',
	'351',
	'352',
	'353',
	'354',
	'355',
	'356',
	'357',
	'358',
	'359',
	'370',
	'371',
	'372',
	'373',
	'374',
	'375',
	'376',
	'377',
	'378',
	'380',
	'381',
	'382',
	'383',
	'385',
	'386',
	'387',
	'389',
	'420',
	'421',
	'423',
	'500',
	'501',
	'502',
	'503',
	'504',
	'505',
	'506',
	'507',
	'508',
	'509',
	'590',
	'591',
	'592',
	'593',
	'594',
	'595',
	'596',
	'597',
	'598',
	'599',
	'670',
	'672',
	'673',
	'674',
	'675',
	'676',
	'677',
	'678',
	'679',
	'680',
	'681',
	'682',
	'683',
	'685',
	'686',
	'687',
	'688',
	'689',
	'690',
	'691',
	'692',
	'800',
	'808',
	'850',
	'852',
	'853',
	'855',
	'856',
	'870',
	'878',
	'880',
	'881',
	'882',
	'883',
	'886',
	'888',
	'960',
	'961',
	'962',
	'963',
	'964',
	'965',
	'966',
	'967',
	'968',
	'970',
	'971',
	'972',
	'973',
	'974',
	'975',
	'976',
	'977',
	'979',
	'992',
	'993',
	'994',
	'995',
	'996',
	'998',
];

interface CountryValues {
	iso: string;
	name: string;
}
export const COUNTRIES: CountryValues[] = [
	{ iso: 'AF', name: 'Afghanistan' },
	{ iso: 'AL', name: 'Albania' },
	{ iso: 'DZ', name: 'Algeria' },
	{ iso: 'AS', name: 'American Samoa' },
	{ iso: 'AD', name: 'Andorra' },
	{ iso: 'AO', name: 'Angola' },
	{ iso: 'AI', name: 'Anguilla' },
	{ iso: 'AQ', name: 'Antarctica' },
	{ iso: 'AG', name: 'Antigua & Barbuda' },
	{ iso: 'AR', name: 'Argentina' },
	{ iso: 'AM', name: 'Armenia' },
	{ iso: 'AW', name: 'Aruba' },
	{ iso: 'AU', name: 'Australia' },
	{ iso: 'AT', name: 'Austria' },
	{ iso: 'AZ', name: 'Azerbaijan' },
	{ iso: 'BS', name: 'Bahamas' },
	{ iso: 'BH', name: 'Bahrain' },
	{ iso: 'BD', name: 'Bangladesh' },
	{ iso: 'BB', name: 'Barbados' },
	{ iso: 'BY', name: 'Belarus' },
	{ iso: 'BE', name: 'Belgium' },
	{ iso: 'BZ', name: 'Belize' },
	{ iso: 'BJ', name: 'Benin' },
	{ iso: 'BM', name: 'Bermuda' },
	{ iso: 'BT', name: 'Bhutan' },
	{ iso: 'BO', name: 'Bolivia' },
	{ iso: 'BQ', name: 'Bonaire, Saint Eustatius and Saba' },
	{ iso: 'BA', name: 'Bosnia-Herzegovina' },
	{ iso: 'BW', name: 'Botswana' },
	{ iso: 'BV', name: 'Bouvet Island' },
	{ iso: 'BR', name: 'Brazil' },
	{ iso: 'IO', name: 'British Indian Ocean Territory' },
	{ iso: 'VG', name: 'British Virgin Islands' },
	{ iso: 'BN', name: 'Brunei Darussalam' },
	{ iso: 'BG', name: 'Bulgaria' },
	{ iso: 'BF', name: 'Burkina Faso' },
	{ iso: 'BI', name: 'Burundi' },
	{ iso: 'KH', name: 'Cambodia' },
	{ iso: 'CM', name: 'Cameroon' },
	{ iso: 'CA', name: 'Canada' },
	{ iso: 'CV', name: 'Cape Verde Islands' },
	{ iso: 'KY', name: 'Cayman Islands' },
	{ iso: 'CF', name: 'Central African Republic' },
	{ iso: 'TD', name: 'Chad' },
	{ iso: 'CL', name: 'Chile' },
	{ iso: 'CN', name: 'China' },
	{ iso: 'CX', name: 'Christmas Island' },
	{ iso: 'CC', name: 'Cocos (Keeling) Islands' },
	{ iso: 'CO', name: 'Colombia' },
	{ iso: 'KM', name: 'Comoros' },
	{ iso: 'CG', name: 'Congo (Brazzaville)' },
	{ iso: 'CD', name: 'Congo (Kinshasa)' },
	{ iso: 'CK', name: 'Cook Islands' },
	{ iso: 'CR', name: 'Costa Rica' },
	{ iso: 'HR', name: 'Croatia' },
	{ iso: 'CU', name: 'Cuba' },
	{ iso: 'CW', name: 'Cura√ßao' },
	{ iso: 'CY', name: 'Cyprus' },
	{ iso: 'CZ', name: 'Czech Republic' },
	{ iso: 'DK', name: 'Denmark' },
	{ iso: 'DJ', name: 'Djibouti' },
	{ iso: 'DM', name: 'Dominica' },
	{ iso: 'DO', name: 'Dominican Republic' },
	{ iso: 'TL', name: 'East Timor' },
	{ iso: 'EC', name: 'Ecuador' },
	{ iso: 'EG', name: 'Egypt' },
	{ iso: 'SV', name: 'El Salvador' },
	{ iso: 'GQ', name: 'Equatorial Guinea' },
	{ iso: 'ER', name: 'Eritrea' },
	{ iso: 'EE', name: 'Estonia' },
	{ iso: 'ET', name: 'Ethiopia' },
	{ iso: 'FK', name: 'Falkland Islands' },
	{ iso: 'FO', name: 'Faroe Islands' },
	{ iso: 'FJ', name: 'Fiji' },
	{ iso: 'FI', name: 'Finland' },
	{ iso: 'FR', name: 'France' },
	{ iso: 'GF', name: 'French Guiana' },
	{ iso: 'PF', name: 'French Polynesia' },
	{ iso: 'TF', name: 'French Southern Territories' },
	{ iso: 'GA', name: 'Gabon' },
	{ iso: 'GM', name: 'Gambia' },
	{ iso: 'GE', name: 'Georgia' },
	{ iso: 'DE', name: 'Germany' },
	{ iso: 'GH', name: 'Ghana' },
	{ iso: 'GI', name: 'Gibraltar' },
	{ iso: 'GR', name: 'Greece' },
	{ iso: 'GL', name: 'Greenland' },
	{ iso: 'GD', name: 'Grenada' },
	{ iso: 'GP', name: 'Guadeloupe' },
	{ iso: 'GU', name: 'Guam' },
	{ iso: 'GT', name: 'Guatemala' },
	{ iso: 'GG', name: 'Guernsey' },
	{ iso: 'GN', name: 'Guinea' },
	{ iso: 'GW', name: 'Guinea-Bissau' },
	{ iso: 'GY', name: 'Guyana' },
	{ iso: 'HT', name: 'Haiti' },
	{ iso: 'HM', name: 'Heard Island and McDonald Islands' },
	{ iso: 'VA', name: 'Holy See' },
	{ iso: 'HN', name: 'Honduras' },
	{ iso: 'HK', name: 'Hong Kong' },
	{ iso: 'HU', name: 'Hungary' },
	{ iso: 'IS', name: 'Iceland' },
	{ iso: 'IN', name: 'India' },
	{ iso: 'ID', name: 'Indonesia' },
	{ iso: 'IR', name: 'Iran' },
	{ iso: 'IQ', name: 'Iraq' },
	{ iso: 'IE', name: 'Ireland' },
	{ iso: 'IM', name: 'Isle of Man' },
	{ iso: 'IL', name: 'Israel' },
	{ iso: 'IT', name: 'Italy' },
	{ iso: 'CI', name: 'Ivory Coast' },
	{ iso: 'JM', name: 'Jamaica' },
	{ iso: 'JP', name: 'Japan' },
	{ iso: 'JE', name: 'Jersey' },
	{ iso: 'JO', name: 'Jordan' },
	{ iso: 'KZ', name: 'Kazakhstan' },
	{ iso: 'KE', name: 'Kenya' },
	{ iso: 'KI', name: 'Kiribati' },
	{ iso: 'KW', name: 'Kuwait' },
	{ iso: 'KG', name: 'Kyrgyzstan' },
	{ iso: 'LA', name: 'Laos' },
	{ iso: 'LV', name: 'Latvia' },
	{ iso: 'LB', name: 'Lebanon' },
	{ iso: 'LS', name: 'Lesotho' },
	{ iso: 'LR', name: 'Liberia' },
	{ iso: 'LY', name: 'Libya' },
	{ iso: 'LI', name: 'Liechtenstein' },
	{ iso: 'LT', name: 'Lithuania' },
	{ iso: 'LU', name: 'Luxembourg' },
	{ iso: 'MO', name: 'Macau' },
	{ iso: 'MK', name: 'Macedonia' },
	{ iso: 'MG', name: 'Madagascar' },
	{ iso: 'MW', name: 'Malawi' },
	{ iso: 'MY', name: 'Malaysia' },
	{ iso: 'MV', name: 'Maldives' },
	{ iso: 'ML', name: 'Mali' },
	{ iso: 'MT', name: 'Malta' },
	{ iso: 'MH', name: 'Marshall Islands' },
	{ iso: 'MQ', name: 'Martinique' },
	{ iso: 'MR', name: 'Mauritania' },
	{ iso: 'MU', name: 'Mauritius' },
	{ iso: 'YT', name: 'Mayotte' },
	{ iso: 'MX', name: 'Mexico' },
	{ iso: 'FM', name: 'Micronesia' },
	{ iso: 'MD', name: 'Moldova' },
	{ iso: 'MC', name: 'Monaco' },
	{ iso: 'MN', name: 'Mongolia' },
	{ iso: 'ME', name: 'Montenegro' },
	{ iso: 'MS', name: 'Montserrat' },
	{ iso: 'MA', name: 'Morocco' },
	{ iso: 'MZ', name: 'Mozambique' },
	{ iso: 'MM', name: 'Myanmar' },
	{ iso: 'NA', name: 'Namibia' },
	{ iso: 'NR', name: 'Nauru' },
	{ iso: 'NP', name: 'Nepal' },
	{ iso: 'NL', name: 'Netherlands' },
	{ iso: 'NC', name: 'New Caledonia' },
	{ iso: 'NZ', name: 'New Zealand' },
	{ iso: 'NI', name: 'Nicaragua' },
	{ iso: 'NE', name: 'Niger' },
	{ iso: 'NG', name: 'Nigeria' },
	{ iso: 'NU', name: 'Niue' },
	{ iso: 'NF', name: 'Norfolk Island' },
	{ iso: 'KP', name: 'North Korea' },
	{ iso: 'MP', name: 'Northern Mariana Islands' },
	{ iso: 'NO', name: 'Norway' },
	{ iso: 'OM', name: 'Oman' },
	{ iso: 'PK', name: 'Pakistan' },
	{ iso: 'PW', name: 'Palau' },
	{ iso: 'PS', name: 'Palestinian Territories' },
	{ iso: 'PA', name: 'Panama' },
	{ iso: 'PG', name: 'Papua New Guinea' },
	{ iso: 'PY', name: 'Paraguay' },
	{ iso: 'PE', name: 'Peru' },
	{ iso: 'PH', name: 'Philippines' },
	{ iso: 'PN', name: 'Pitcairn Islands' },
	{ iso: 'PL', name: 'Poland' },
	{ iso: 'PT', name: 'Portugal' },
	{ iso: 'PR', name: 'Puerto Rico' },
	{ iso: 'QA', name: 'Qatar' },
	{ iso: 'RO', name: 'Romania' },
	{ iso: 'RU', name: 'Russia' },
	{ iso: 'RW', name: 'Rwanda' },
	{ iso: 'RE', name: 'R√©union' },
	{ iso: 'BL', name: 'Saint Barth√©lemy' },
	{ iso: 'KN', name: 'Saint Christopher & Nevis' },
	{ iso: 'SH', name: 'Saint Helena' },
	{ iso: 'LC', name: 'Saint Lucia' },
	{ iso: 'MF', name: 'Saint Martin' },
	{ iso: 'PM', name: 'Saint Pierre & Miquelon' },
	{ iso: 'VC', name: 'Saint Vincent & The Grenadines' },
	{ iso: 'WS', name: 'Samoa' },
	{ iso: 'SM', name: 'San Marino' },
	{ iso: 'ST', name: 'Sao Tome & Principe' },
	{ iso: 'SA', name: 'Saudi Arabia' },
	{ iso: 'SN', name: 'Senegal' },
	{ iso: 'RS', name: 'Serbia' },
	{ iso: 'SC', name: 'Seychelles' },
	{ iso: 'SL', name: 'Sierra Leone' },
	{ iso: 'SG', name: 'Singapore' },
	{ iso: 'SX', name: 'Sint Maarten' },
	{ iso: 'SK', name: 'Slovakia' },
	{ iso: 'SI', name: 'Slovenia' },
	{ iso: 'SB', name: 'Solomon Islands' },
	{ iso: 'SO', name: 'Somalia' },
	{ iso: 'ZA', name: 'South Africa' },
	{ iso: 'GS', name: 'South Georgia & The South Sandwich Islands' },
	{ iso: 'KR', name: 'South Korea' },
	{ iso: 'SS', name: 'South Sudan' },
	{ iso: 'ES', name: 'Spain' },
	{ iso: 'LK', name: 'Sri Lanka' },
	{ iso: 'SD', name: 'Sudan' },
	{ iso: 'SR', name: 'Suriname' },
	{ iso: 'SJ', name: 'Svalbard and Jan Mayen' },
	{ iso: 'SZ', name: 'Swaziland' },
	{ iso: 'SE', name: 'Sweden' },
	{ iso: 'CH', name: 'Switzerland' },
	{ iso: 'SY', name: 'Syria' },
	{ iso: 'TW', name: 'Taiwan' },
	{ iso: 'TJ', name: 'Tajikistan' },
	{ iso: 'TZ', name: 'Tanzania' },
	{ iso: 'TH', name: 'Thailand' },
	{ iso: 'TG', name: 'Togo' },
	{ iso: 'TK', name: 'Tokelau' },
	{ iso: 'TO', name: 'Tonga' },
	{ iso: 'TT', name: 'Trinidad & Tobago' },
	{ iso: 'TN', name: 'Tunisia' },
	{ iso: 'TR', name: 'Turkey' },
	{ iso: 'TM', name: 'Turkmenistan' },
	{ iso: 'TC', name: 'Turks & Caicos Islands' },
	{ iso: 'TV', name: 'Tuvalu' },
	{ iso: 'UG', name: 'Uganda' },
	{ iso: 'UA', name: 'Ukraine' },
	{ iso: 'AE', name: 'United Arab Emirates' },
	{ iso: 'GB', name: 'United Kingdom' },
	{ iso: 'US', name: 'United States' },
	{ iso: 'UM', name: 'United States Minor Outlying Islands' },
	{ iso: 'VI', name: 'United States Virgin Islands' },
	{ iso: 'UY', name: 'Uruguay' },
	{ iso: 'UZ', name: 'Uzbekistan' },
	{ iso: 'VU', name: 'Vanuatu' },
	{ iso: 'VE', name: 'Venezuela' },
	{ iso: 'VN', name: 'Vietnam' },
	{ iso: 'WF', name: 'Wallis & Futuna' },
	{ iso: 'EH', name: 'Western Sahara' },
	{ iso: 'YE', name: 'Yemen' },
	{ iso: 'ZM', name: 'Zambia' },
	{ iso: 'ZW', name: 'Zimbabwe' },
	{ iso: 'AX', name: '√Öland Islands' },
];
