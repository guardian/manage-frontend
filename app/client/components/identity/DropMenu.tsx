import { CSSObject } from '@emotion/core';
import { FC, useState } from 'react';
import palette from '../../colours';
import { serif } from '../../styles/fonts';
interface DropMenuProps {
	title: string;
	color: string;
}

const styles = {
	root: {
		borderColor: palette.neutral['4'],
		borderTop: `1px solid ${palette.neutral['6']}`,
		cursor: 'pointer',
		padding: '3px 0 12px',
		position: 'relative',
	} as CSSObject,
	header: {
		fontSize: '17px',
		fontFamily: serif,
		fontWeight: 'bold',
		lineHeight: '24px',
		textTransform: 'capitalize',
		':after': {
			content: "''",
			border: '2px solid currentColor',
			borderLeft: 'transparent',
			borderTop: 'transparent',
			boxSizing: 'content-box',
			display: 'inline-block',
			height: '5px',
			color: 'inherit',
			transform: 'translateY(-2px) rotate(45deg)',
			transition: 'transform 250ms ease-out',
			verticalAlign: 'middle',
			width: '5px',
			marginLeft: '4px',
		},
		'&.open:after': {
			transform: 'rotate(225deg)',
		},
	} as CSSObject,
};

export const DropMenu: FC<DropMenuProps> = (props) => {
	const { children, color, title } = props;
	const [open, setOpen] = useState(false);
	return (
		<div css={styles.root}>
			<div
				className={open ? 'open' : undefined}
				css={{ ...styles.header, color }}
				onClick={() => setOpen(!open)}
			>
				{title}
			</div>
			{open && children}
		</div>
	);
};
