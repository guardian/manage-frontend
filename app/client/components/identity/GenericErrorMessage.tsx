import { forwardRef } from 'react';
import palette from '../../colours';
import { errorMessageCss } from './sharedStyles';

export type GenericErrorMessageRef = HTMLDivElement;
export const GenericErrorMessage = forwardRef<GenericErrorMessageRef, {}>(
	(_, ref) => {
		return (
			<div ref={ref} css={errorMessageCss}>
				<p css={{ marginBottom: '10px' }}>
					Sorry, something went wrong!
				</p>
				<p css={{ marginBottom: '0' }}>
					<a
						css={{
							color: palette.blue.dark,
							cursor: 'pointer',
						}}
						onClick={() => window.location.reload()}
					>
						Refresh this page
					</a>
				</p>
			</div>
		);
	},
);
