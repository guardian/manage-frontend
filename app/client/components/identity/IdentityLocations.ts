import { conf } from '../../../server/config';

const url = (subdomain: string, domain: string, path?: string) =>
	`https://${subdomain}.${domain}${path ? path : ''}`;

const DOMAIN =
	typeof window !== 'undefined' && window.guardian
		? window.guardian.domain
		: conf.DOMAIN;

const IMAGE_DOMAIN =
	DOMAIN === 'theguardian.com' ? 'guim.co.uk' : 'guimcode.co.uk';

const IDAPI_URL =
	DOMAIN === 'thegulocal.com' ? '/idapicodeproxy' : url('idapi', DOMAIN);

const AVATAR_URL =
	DOMAIN === 'thegulocal.com' ? '/avatarcodeproxy' : url('avatar', DOMAIN);

const getIdentityLocations = (domain: string) => ({
	COMMUNITY_FAQS: url('www', domain, '/community-faqs'),
	CONTACT_AND_DELIVERY_HELP: url(
		'manage',
		'theguardian.com',
		'/help-centre/article/i-need-to-change-my-delivery-address',
	),
	CHANGE_EMAIL: url('profile', domain, '/account/edit'),
	RESET_PASSWORD: url('profile', domain, '/reset'),
	MANAGE_JOB_ALERTS: url('jobs', domain, '/your-jobs/?ActiveSection=JbeList'),
	VERIFY_EMAIL: url('profile', domain, '/verify-email'),
	IDAPI: IDAPI_URL,
	AVATAR: AVATAR_URL,
	AVATAR_USER_IMAGES: url('avatar', IMAGE_DOMAIN, '/user'),
	DELETE_ACCOUNT: url('profile', domain, '/delete'),
});

export const IdentityLocations = getIdentityLocations(DOMAIN);
