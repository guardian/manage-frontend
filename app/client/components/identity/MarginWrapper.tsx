import { FC } from "react";
import { minWidth } from "../../styles/breakpoints";

export const MarginWrapper: FC<{}> = ({ children }) => {
  return (
    <div
      css={{
        boxSizing: "border-box",
        margin: "0 auto",
        maxWidth: "620px",
        [minWidth.desktop]: {
          maxWidth: "100%",
        },
      }}
    >
      {children}
    </div>
  );
};
