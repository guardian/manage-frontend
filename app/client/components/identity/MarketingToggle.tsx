import React, { FC } from "react";
import { sans } from "../../styles/fonts";
import { css } from "@emotion/core";
import { ToggleSwitch } from "./ToggleSwitch";

interface MarketingToggleProps {
  id: string;
  description?: string;
  title?: string;
  selected?: boolean;
  onClick: (id: string) => {};
}

const standardText = {
  fontFamily: sans,
  fontSize: "14px",
  lineHeight: "1.333"
};

const getDescription = (description: MarketingToggleProps["description"]) => (
  <p
    css={[
      standardText,
      {
        margin: "0",
        padding: "2.88px 90px 0 0"
      }
    ]}
  >
    {description}
  </p>
);

export const MarketingToggle: FC<MarketingToggleProps> = (props) => {
  const { id, description, selected, title, onClick } = props;
  return (
    <div
      onClick={(e) => {
        // Checkboxes inside labels will trigger click events twice.
        // Ignore the input click event
        if (e.target instanceof Element && e.target.nodeName === "INPUT") {
          return;
        }
        onClick(id);
      }}
      css={[
        standardText,
        {
          marginTop: "12px",
          position: "relative"
        }
      ]}
    >
      <div css={{ left: 0 }}>
        <ToggleSwitch
          cssOverrides={css`
            label {
              ${standardText};
              padding-right: 5px;
              font-weight: bold;
            }
          `}
          label={title}
          labelPosition="left"
          labelId={id}
          checked={!!selected}
          onClick={(_) => {
            return;
          }}
        />
      </div>
      {description && getDescription(description)}
    </div>
  );
};
