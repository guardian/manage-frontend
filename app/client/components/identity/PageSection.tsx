import { FC, ReactNode } from 'react';
import palette from '../../colours';
import { minWidth } from '../../styles/breakpoints';
import { sans } from '../../styles/fonts';

interface PageSectionProps {
	title?: string;
	description?: string | ReactNode;
	subtext?: string;
}

const getTitle = (title: PageSectionProps['title']) => (
	<h2
		css={{
			fontSize: '17px',
			lineHeight: '24px',
			fontWeight: 'bold',
			margin: '0 0 6px 0',
		}}
	>
		{title}
	</h2>
);

const getDescription = (description: PageSectionProps['description']) => (
	<p
		css={{
			fontFamily: sans,
			fontSize: '14px',
			marginBottom: '8px',
		}}
	>
		{description}
	</p>
);

const getSubtext = (subtext: PageSectionProps['subtext']) => (
	<p
		css={{
			fontSize: '13px',
			lineHeight: '18px',
			fontFamily: sans,
			color: palette.neutral['3'],
			marginBottom: '12px',
			marginTop: '0',
		}}
	>
		{subtext}
	</p>
);

export const PageSection: FC<PageSectionProps> = (props) => {
	const { children, description, title, subtext } = props;
	return (
		<div
			css={{
				[minWidth.desktop]: {
					display: 'flex',
				},
			}}
		>
			<div
				css={{
					[minWidth.desktop]: {
						paddingRight: '100px',
						boxSizing: 'content-box',
						flexBasis: '220px',
						minWidth: '220px',
					},
				}}
			>
				{title && getTitle(title)}
				{description && getDescription(description)}
				{subtext && getSubtext(subtext)}
			</div>
			<div
				css={{
					[minWidth.desktop]: {
						maxWidth: '460px',
						flexGrow: 1,
					},
				}}
			>
				{children}
			</div>
		</div>
	);
};
