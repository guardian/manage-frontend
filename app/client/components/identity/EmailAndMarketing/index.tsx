import React, { useEffect, useState } from "react";
import { ProductDetail } from "../../../../shared/productResponse";
import { GROUPED_PRODUCT_TYPES } from "../../../../shared/productTypes";
import { allProductsDetailFetcher } from "../../../productUtils";
import { NAV_LINKS } from "../../nav/navConfig";
import { PageContainer } from "../../page";
import { Spinner } from "../../spinner";
import { WithStandardTopMargin } from "../../WithStandardTopMargin";
import {
  GenericErrorMessage,
  GenericErrorMessageRef
} from "../GenericErrorMessage";
import { ConsentOptions, Users } from "../identity";
import { IdentityLocations } from "../IdentityLocations";
import { Lines } from "../Lines";
import { ConsentOption } from "../models";
import { Actions, useConsentOptions } from "../useConsentOptions";
import { ConsentSection } from "./ConsentSection";
import { EmailSettingsSection } from "./EmailSettingsSection";
import { NewsletterSection } from "./NewsletterSection";
import { OptOutSection } from "./OptOutSection";

const EmailAndMarketing = (_: { path?: string }) => {
  const { options, error, subscribe, unsubscribe, unsubscribeAll } = Actions;
  const [email, setEmail] = useState<string>("");
  const [removed, setRemoved] = useState(false);
  const [state, dispatch] = useConsentOptions();

  const toggleSubscription = async (id: string) => {
    const option = ConsentOptions.findById(state.options, id);
    try {
      if (option === undefined) {
        throw Error("Id not found");
      }
      if (option.subscribed) {
        await ConsentOptions.unsubscribe(option);
        dispatch(unsubscribe(id));
      } else {
        await ConsentOptions.subscribe(option);
        dispatch(subscribe(id));
      }
    } catch (e) {
      dispatch(error(e));
    }
  };

  const setRemoveAllEmailConsents = async () => {
    try {
      await ConsentOptions.unsubscribeAll();
      setRemoved(true);
      dispatch(unsubscribeAll());
    } catch (e) {
      dispatch(error(e));
    }
  };

  useEffect(() => {
    const makeInitialAPICalls = async () => {
      try {
        const user = await Users.getCurrentUser();
        if (!user.validated) {
          window.location.assign(IdentityLocations.VERIFY_EMAIL);
          return;
        }
        const productDetailsResponse = await allProductsDetailFetcher();
        const productDetails: ProductDetail[] = await productDetailsResponse.json();
        const consentOptions = await ConsentOptions.getAll();
        const consentsWithFilteredSoftOptIns = consentOptions.filter(
          (consent: ConsentOption) =>
            consent.isProduct
              ? productDetails.some(productDetail => {
                  const groupedProductType =
                    GROUPED_PRODUCT_TYPES[productDetail.mmaCategory];
                  const specificProductType = groupedProductType.mapGroupedToSpecific(
                    productDetail
                  );
                  return specificProductType.softOptInIDs.includes(consent.id);
                })
              : true
        );

        setEmail(user.primaryEmailAddress);
        dispatch(options(consentsWithFilteredSoftOptIns));
      } catch (e) {
        dispatch(error(e));
      }
    };
    makeInitialAPICalls();
  }, []);

  const newsletters = ConsentOptions.newsletters(state.options);
  const consents = ConsentOptions.consents(state.options);
  const loading = newsletters.length === 0 && consents.length === 0;

  const errorRef = React.createRef<GenericErrorMessageRef>();

  useEffect(() => {
    if (state.error && errorRef.current) {
      window.scrollTo(0, errorRef.current.offsetTop - 20);
    }
  }, [state.error]);

  const errorMessage = (
    <WithStandardTopMargin>
      <GenericErrorMessage ref={errorRef} />
    </WithStandardTopMargin>
  );

  const content = (
    <>
      <WithStandardTopMargin>
        <NewsletterSection
          newsletters={newsletters}
          clickHandler={toggleSubscription}
        />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <Lines n={4} />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <ConsentSection consents={consents} clickHandler={toggleSubscription} />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <Lines n={1} />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <EmailSettingsSection
          email={email}
          actionHandler={setRemoveAllEmailConsents}
          removed={removed}
        />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <Lines n={4} />
      </WithStandardTopMargin>
      <WithStandardTopMargin>
        <OptOutSection consents={consents} clickHandler={toggleSubscription} />
      </WithStandardTopMargin>
    </>
  );

  const loader = (
    <WithStandardTopMargin>
      <Spinner loadingMessage="Loading your email preferences..." />
    </WithStandardTopMargin>
  );

  return (
    <PageContainer
      selectedNavItem={NAV_LINKS.emailPrefs}
      pageTitle="Emails &amp; marketing"
    >
      {state.error ? errorMessage : null}
      {loading ? (!state.error ? loader : null) : content}
    </PageContainer>
  );
};

export default EmailAndMarketing;
