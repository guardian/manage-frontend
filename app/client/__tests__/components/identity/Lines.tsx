import serializer from 'jest-emotion';
import React from 'react';
import { create } from 'react-test-renderer';
import { Lines } from '../../../components/identity/Lines';

expect.addSnapshotSerializer(serializer);

describe('Lines', () => {
	it('draws one line when n = 1', () => {
		const container = create(<Lines n={1} />);
		expect(container).toMatchSnapshot();
	});
	it('draws one line when n = 2', () => {
		const container = create(<Lines n={2} />);
		expect(container).toMatchSnapshot();
	});
});
