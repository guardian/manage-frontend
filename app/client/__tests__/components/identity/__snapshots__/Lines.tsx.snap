// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lines draws one line when n = 1 1`] = `
.emotion-0 {
  background-image: repeating-linear-gradient(to bottom, #dcdcdc, #dcdcdc 1px, transparent 1px, transparent 4px);
  background-repeat: repeat;
  -webkit-background-position: top;
  background-position: top;
  height: 4px;
  border: 0;
  margin: 12px auto 6px;
}

<hr
  className="emotion-0"
/>
`;

exports[`Lines draws one line when n = 2 1`] = `
.emotion-0 {
  background-image: repeating-linear-gradient(to bottom, #dcdcdc, #dcdcdc 1px, transparent 1px, transparent 4px);
  background-repeat: repeat;
  -webkit-background-position: top;
  background-position: top;
  height: 8px;
  border: 0;
  margin: 12px auto 6px;
}

<hr
  className="emotion-0"
/>
`;
